libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+/**
+ * @module botbuilder-ai
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Activity, ActivityTypes } from 'botbuilder-core';
+import { WaterfallDialog, Dialog, DialogTurnResult, DialogContext, WaterfallStepContext, DialogTurnStatus, DialogReason } from 'botbuilder-dialogs';
+import { QnAMakerOptions } from './qnamaker-interfaces/qnamakerOptions';
+import { RankerTypes } from './qnamaker-interfaces/rankerTypes';
+import { QnAMaker, QnAMakerResult } from './';
+import { FeedbackRecord, FeedbackRecords, QnAMakerMetadata } from './qnamaker-interfaces';
+import { QnACardBuilder } from './qnaCardBuilder';
+
+/**
+ * QnAMakerDialog response options.
+ */
+export interface QnAMakerDialogResponseOptions {
+    /**
+    * Title for active learning card.
+    */
+    activeLearningCardTitle: string;
+    /**
+    * Text shown for 'no match' option on active learning card.
+    */
+    cardNoMatchText: string;
+    /**
+    * Activity to be sent in the event of no answer found in KB.
+    */
+    noAnswer: Activity;
+    /**
+    * Activity to be sent in the end that the 'no match' option is selected on active learning card.
+    */
+    cardNoMatchResponse: Activity;
+}
+
+/**
+ * Options for QnAMakerDialog.
+ */
+export interface QnAMakerDialogOptions {
+    /**
+    * Options for QnAMaker knowledgebase.
+    */
+    qnaMakerOptions: QnAMakerOptions;
+    /**
+    * QnAMakerDialog response options.
+    */
+    qnaDialogResponseOptions: QnAMakerDialogResponseOptions;
+}
+
+/**
+ * Query QnAMaker knowledgebase using user utterance. The dialog will also present user with appropriate multi-turn prompt or active learning options.
+ */
+export class QnAMakerDialog extends WaterfallDialog {
```

Comment:
>WaterfallDialog [](start = 36, length = 15)

This doesn't feel like it should derive from waterfall... The fact you need to loop, conditionally run steps, etc. means that you need your own state machine. The waterfall is just complicating your life here.

Reply:
@Stevenic I agree - I think right now though, if we want to maintain parity we need to live with this as a waterfall given the C# version has already been released https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs